### PR TITLE
provider: Stabilization of remaining data source id attributes and enable R015, R016, R017 linters

### LINKED ISSUE #14579: Data sources with non-deterministic IDs show perpetual diff with Terraform 0.13 (R015, R016, R017 linters)
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform CLI and Terraform AWS Provider Version

Terraform v0.13.0
AWS Provider v3.1.0

### Affected Resource(s)

```console
$ grep 'd.SetId' aws/data_source*.go | grep Now
aws/data_source_aws_acm_certificate.go:	d.SetId(time.Now().UTC().String())
aws/data_source_aws_autoscaling_group.go:	d.SetId(time.Now().UTC().String())
aws/data_source_aws_autoscaling_groups.go:	d.SetId(time.Now().UTC().String())
aws/data_source_aws_availability_zones.go:	d.SetId(time.Now().UTC().String())
aws/data_source_aws_caller_identity.go:	d.SetId(time.Now().UTC().String())
aws/data_source_aws_ebs_default_kms_key.go:	d.SetId(time.Now().UTC().String())
aws/data_source_aws_ebs_encryption_by_default.go:	d.SetId(time.Now().UTC().String())
aws/data_source_aws_ec2_coip_pools.go:	d.SetId(time.Now().UTC().String())
aws/data_source_aws_ec2_local_gateway_route_tables.go:	d.SetId(time.Now().UTC().String())
aws/data_source_aws_ec2_local_gateways.go:	d.SetId(time.Now().UTC().String())
aws/data_source_aws_ecr_authorization_token.go:	d.SetId(time.Now().UTC().String())
aws/data_source_aws_ecr_image.go:	d.SetId(time.Now().UTC().String())
aws/data_source_aws_efs_access_points.go:	d.SetId(time.Now().UTC().String())
aws/data_source_aws_eks_cluster_auth.go:	d.SetId(time.Now().UTC().String())
aws/data_source_aws_glue_script.go:	d.SetId(time.Now().UTC().String())
aws/data_source_aws_iam_account_alias.go:	d.SetId(time.Now().UTC().String())
aws/data_source_aws_inspector_rules_packages.go:	d.SetId(time.Now().UTC().String())
aws/data_source_aws_kms_alias.go:	d.SetId(time.Now().UTC().String())
aws/data_source_aws_kms_ciphertext.go:	d.SetId(time.Now().UTC().String())
aws/data_source_aws_kms_secrets.go:	d.SetId(time.Now().UTC().String())
aws/data_source_aws_partition.go:	d.SetId(time.Now().UTC().String())
aws/data_source_aws_regions.go:	d.SetId(time.Now().UTC().String())
aws/data_source_aws_route53_resolver_rules.go:	d.SetId(time.Now().UTC().String())
aws/data_source_aws_sns.go:	d.SetId(time.Now().UTC().String())
aws/data_source_aws_vpcs.go:	d.SetId(time.Now().UTC().String())
$ grep 'd.SetId' aws/data_source*.go | grep Unique
aws/data_source_aws_db_event_categories.go:	d.SetId(resource.UniqueId())
aws/data_source_aws_ebs_volumes.go:	d.SetId(resource.UniqueId())
aws/data_source_aws_ec2_instance_type_offering.go:	d.SetId(resource.UniqueId())
aws/data_source_aws_ec2_instance_type_offerings.go:	d.SetId(resource.UniqueId())
aws/data_source_aws_ec2_local_gateway_virtual_interface_groups.go:	d.SetId(resource.UniqueId())
aws/data_source_aws_ec2_spot_price.go:	d.SetId(resource.UniqueId())
aws/data_source_aws_instances.go:	d.SetId(resource.UniqueId())
aws/data_source_aws_network_acls.go:	d.SetId(resource.UniqueId())
aws/data_source_aws_network_interfaces.go:	d.SetId(resource.UniqueId())
aws/data_source_aws_organizations_organizational_units.go:	d.SetId(resource.UniqueId())
aws/data_source_aws_outposts_outposts.go:	d.SetId(resource.UniqueId())
aws/data_source_aws_outposts_sites.go:	d.SetId(resource.UniqueId())
aws/data_source_aws_route_tables.go:	d.SetId(resource.UniqueId())
aws/data_source_aws_s3_bucket_objects.go:	d.SetId(resource.UniqueId())
aws/data_source_aws_security_groups.go:	d.SetId(resource.UniqueId())
```

### Actual Behavior

Data sources whose ID is set to a non-deterministic value (e.g. the current timestamp or a  unique ID) that changes every time a refresh (the data source's Read method) runs show a perpetual diff starting with Terraform 0.13.0.

For example

```hcl
provider "aws" {
  region = "us-west-2"
}

data "aws_availability_zones" "a" {}
```

##### Terraform 0.12

```console
$ terraform apply
data.aws_availability_zones.a: Refreshing state...

Apply complete! Resources: 0 added, 0 changed, 0 destroyed.
$ terraform plan
Refreshing Terraform state in-memory prior to plan...
The refreshed state will be used to calculate this plan, but will not be
persisted to local or remote state storage.

data.aws_availability_zones.a: Refreshing state...

------------------------------------------------------------------------

No changes. Infrastructure is up-to-date.

This means that Terraform did not detect any differences between your
configuration and real physical resources that exist. As a result, no
actions need to be performed.
```

##### Terraform 0.13

```console
$ terraform apply
data.aws_availability_zones.a: Refreshing state...

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
 <= read (data resources)

Terraform will perform the following actions:

  # data.aws_availability_zones.a will be read during apply
  # (config refers to values not yet known)
 <= data "aws_availability_zones" "a"  {
        group_names = [
            "us-west-2",
            "us-west-2-lax-1",
        ]
      ~ id          = "2020-08-11 20:23:08.230813894 +0000 UTC" -> "2020-08-11 20:23:09.61125009 +0000 UTC"
        names       = [
            "us-west-2-lax-1a",
            "us-west-2a",
            "us-west-2b",
            "us-west-2c",
            "us-west-2d",
        ]
        zone_ids    = [
            "usw2-lax1-az1",
            "usw2-az1",
            "usw2-az2",
            "usw2-az3",
            "usw2-az4",
        ]
    }

Plan: 0 to add, 0 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

data.aws_availability_zones.a: Reading... [id=2020-08-11 20:23:08.230813894 +0000 UTC]
data.aws_availability_zones.a: Read complete after 0s [id=2020-08-11 20:23:09.61125009 +0000 UTC]

Apply complete! Resources: 0 added, 0 changed, 0 destroyed.
$ terraform plan
Refreshing Terraform state in-memory prior to plan...
The refreshed state will be used to calculate this plan, but will not be
persisted to local or remote state storage.

data.aws_availability_zones.a: Refreshing state... [id=2020-08-11 20:23:09.61125009 +0000 UTC]

------------------------------------------------------------------------

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
 <= read (data resources)

Terraform will perform the following actions:

  # data.aws_availability_zones.a will be read during apply
  # (config refers to values not yet known)
 <= data "aws_availability_zones" "a"  {
        group_names = [
            "us-west-2",
            "us-west-2-lax-1",
        ]
      ~ id          = "2020-08-11 20:23:19.963587604 +0000 UTC" -> "2020-08-11 20:23:21.360002629 +0000 UTC"
        names       = [
            "us-west-2-lax-1a",
            "us-west-2a",
            "us-west-2b",
            "us-west-2c",
            "us-west-2d",
        ]
        zone_ids    = [
            "usw2-lax1-az1",
            "usw2-az1",
            "usw2-az2",
            "usw2-az3",
            "usw2-az4",
        ]
    }

Plan: 0 to add, 0 to change, 0 to destroy.

------------------------------------------------------------------------
```

### References

Related:

* https://github.com/terraform-providers/terraform-provider-aws/issues/14564
* https://github.com/terraform-providers/terraform-provider-aws/issues/14546

### PR DESCRIPTION
**Pull request by bflad**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Closes #14579
Reference: https://github.com/terraform-providers/terraform-provider-aws/pull/15399

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
NOTES

* data-source/aws_autoscaling_groups: The `id` attribute has changed to the name of the AWS Region. The first apply of this updated data source may show this difference.
* data-source/aws_caller_identity: The `id` attribute has changed to the ID of the AWS Account. The first apply of this updated data source may show this difference.
* data-source/aws_ebs_snapshot_ids: The `id` attribute has changed to the name of the AWS Region. The first apply of this updated data source may show this difference.
* data-source/aws_ebs_volumes: The `id` attribute has changed to the name of the AWS Region. The first apply of this updated data source may show this difference.
* data-source/aws_ec2_coip_pools: The `id` attribute has changed to the name of the AWS Region. The first apply of this updated data source may show this difference.
* data-source/aws_ec2_instance_type_offerings: The `id` attribute has changed to the name of the AWS Region. The first apply of this updated data source may show this difference.
* data-source/aws_ec2_local_gateway_route_tables: The `id` attribute has changed to the name of the AWS Region. The first apply of this updated data source may show this difference.
* data-source/aws_ec2_local_gateway_virtual_interface_groups: The `id` attribute has changed to the name of the AWS Region. The first apply of this updated data source may show this difference.
* data-source/aws_ec2_local_gateways: The `id` attribute has changed to the name of the AWS Region. The first apply of this updated data source may show this difference.
* data-source/aws_ec2_spot_price: The `id` attribute has changed to the name of the AWS Region. The first apply of this updated data source may show this difference.
* data-source/aws_efs_access_points: The `id` attribute has changed to the EFS File System identifier. The first apply of this updated data source may show this difference.
* data-source/aws_glue_script: The `id` attribute has changed to the name of the AWS Region. The first apply of this updated data source may show this difference.
* data-source/aws_inspector_rules_packages: The `id` attribute has changed to the name of the AWS Region. The first apply of this updated data source may show this difference.
* data-source/aws_instances: The `id` attribute has changed to the name of the AWS Region. The first apply of this updated data source may show this difference.
* data-source/aws_kms_ciphertext: The `id` attribute has changed to the KMS Key. The first apply of this updated data source may show this difference.
* data-source/aws_network_acls: The `id` attribute has changed to the name of the AWS Region. The first apply of this updated data source may show this difference.
* data-source/aws_network_interfaces: The `id` attribute has changed to the name of the AWS Region. The first apply of this updated data source may show this difference.
* data-source/aws_organizations_organizational_units: The `id` attribute has changed to the parent identifier. The first apply of this updated data source may show this difference.
* data-source/aws_outposts_outposts: The `id` attribute has changed to the name of the AWS Region. The first apply of this updated data source may show this difference.
* data-source/aws_outposts_sites: The `id` attribute has changed to the name of the AWS Region. The first apply of this updated data source may show this difference.
* data-source/aws_route_tables: The `id` attribute has changed to the name of the AWS Region. The first apply of this updated data source may show this difference.
* data-source/aws_route53_resolver_rules: The `id` attribute has changed to the name of the AWS Region. The first apply of this updated data source may show this difference.
* data-source/aws_s3_bucket_objects: The `id` attribute has changed to the name of the S3 Bucket. The first apply of this updated data source may show this difference.
* data-source/aws_security_groups: The `id` attribute has changed to the name of the AWS Region. The first apply of this updated data source may show this difference.
* data-source/aws_vpc_peering_connections: The `id` attribute has changed to the name of the AWS Region. The first apply of this updated data source may show this difference.
* data-source/aws_vpcs: The `id` attribute has changed to the name of the AWS Region. The first apply of this updated data source may show this difference.

BUG FIXES

* data-source/aws_autoscaling_groups: Prevent plan differences with the `id` attribute
* data-source/aws_caller_identity: Prevent plan differences with the `id` attribute
* data-source/aws_ebs_snapshot_ids: Prevent plan differences with the `id` attribute
* data-source/aws_ebs_volumes: Prevent plan differences with the `id` attribute
* data-source/aws_ec2_coip_pools: Prevent plan differences with the `id` attribute
* data-source/aws_ec2_instance_type_offerings: Prevent plan differences with the `id` attribute
* data-source/aws_ec2_local_gateway_route_tables: Prevent plan differences with the `id` attribute
* data-source/aws_ec2_local_gateway_virtual_interface_groups: Prevent plan differences with the `id` attribute
* data-source/aws_ec2_local_gateways: Prevent plan differences with the `id` attribute
* data-source/aws_ec2_spot_price: Prevent plan differences with the `id` attribute
* data-source/aws_efs_access_points: Prevent plan differences with the `id` attribute
* data-source/aws_glue_script: Prevent plan differences with the `id` attribute
* data-source/aws_inspector_rules_packages: Prevent plan differences with the `id` attribute
* data-source/aws_instances: Prevent plan differences with the `id` attribute
* data-source/aws_kms_ciphertext: Prevent plan differences with the `id` attribute
* data-source/aws_network_acls: Prevent plan differences with the `id` attribute
* data-source/aws_network_interfaces: Prevent plan differences with the `id` attribute
* data-source/aws_organizations_organizational_units: Prevent plan differences with the `id` attribute
* data-source/aws_outposts_outposts: Prevent plan differences with the `id` attribute
* data-source/aws_outposts_sites: Prevent plan differences with the `id` attribute
* data-source/aws_route_tables: Prevent plan differences with the `id` attribute
* data-source/aws_route53_resolver_rules: Prevent plan differences with the `id` attribute
* data-source/aws_s3_bucket_objects: Prevent plan differences with the `id` attribute
* data-source/aws_security_groups: Prevent plan differences with the `id` attribute
* data-source/aws_vpc_peering_connections: Prevent plan differences with the `id` attribute
* data-source/aws_vpcs: Prevent plan differences with the `id` attribute
```

Output from acceptance testing:

```
--- PASS: TestAccAWSAutoscalingAttachment_albTargetGroup (111.36s)
--- PASS: TestAccAWSAutoscalingAttachment_elb (118.66s)

--- PASS: TestAccAWSAutoscalingGroups_basic (162.22s)

--- PASS: TestAccAWSCallerIdentity_basic (55.69s)

--- PASS: TestAccAWSEBSEncryptionByDefault_basic (52.71s)

--- PASS: TestAccAWSEc2InstanceTypeOfferingsDataSource_Filter (57.28s)
--- PASS: TestAccAWSEc2InstanceTypeOfferingsDataSource_LocationType (58.79s)

--- PASS: TestAccAwsEc2SpotPriceDataSource_basic (54.47s)
--- PASS: TestAccAwsEc2SpotPriceDataSource_Filter (59.40s)

--- PASS: TestAccAWSELBAttachment_basic (185.93s)
--- PASS: TestAccAWSELBAttachment_drift (113.04s)

--- PASS: TestAccAWSInspectorRulesPackages_basic (58.00s)

--- PASS: TestAccAWSInstancesDataSource_basic (137.19s)
--- PASS: TestAccAWSInstancesDataSource_instance_state_names (122.56s)
--- PASS: TestAccAWSInstancesDataSource_tags (120.98s)

--- PASS: TestAccDataSourceAwsEbsSnapshotIds_basic (92.82s)
--- PASS: TestAccDataSourceAwsEbsSnapshotIds_empty (57.58s)
--- PASS: TestAccDataSourceAwsEbsSnapshotIds_sorted (195.34s)

--- PASS: TestAccDataSourceAwsEbsVolumes_basic (178.56s)

--- PASS: TestAccDataSourceAWSEFSAccessPoints_basic (71.87s)

--- PASS: TestAccDataSourceAWSGlueScript_Language_Python (57.14s)
--- PASS: TestAccDataSourceAWSGlueScript_Language_Scala (57.67s)

--- PASS: TestAccDataSourceAwsKmsCiphertext_basic (61.22s)
--- PASS: TestAccDataSourceAwsKmsCiphertext_validate (60.99s)
--- PASS: TestAccDataSourceAwsKmsCiphertext_validate_withContext (60.77s)

--- PASS: TestAccDataSourceAwsNetworkAcls_basic (112.54s)
--- PASS: TestAccDataSourceAwsNetworkAcls_Filter (62.55s)
--- PASS: TestAccDataSourceAwsNetworkAcls_Tags (58.95s)
--- PASS: TestAccDataSourceAwsNetworkAcls_VpcID (60.22s)

--- PASS: TestAccDataSourceAwsNetworkInterfaces_Filter (95.49s)
--- PASS: TestAccDataSourceAwsNetworkInterfaces_Tags (97.05s)

--- PASS: TestAccDataSourceAwsRoute53ResolverRules_basic (57.40s)
--- PASS: TestAccDataSourceAwsRoute53ResolverRules_ResolverEndpointId (264.98s)

--- PASS: TestAccDataSourceAwsRouteTables_basic (120.92s)

--- PASS: TestAccDataSourceAWSS3BucketObjects_all (120.73s)
--- PASS: TestAccDataSourceAWSS3BucketObjects_basic (122.61s)
--- PASS: TestAccDataSourceAWSS3BucketObjects_basicViaAccessPoint (122.87s)
--- PASS: TestAccDataSourceAWSS3BucketObjects_encoded (115.55s)
--- PASS: TestAccDataSourceAWSS3BucketObjects_fetchOwner (102.52s)
--- PASS: TestAccDataSourceAWSS3BucketObjects_maxKeys (100.50s)
--- PASS: TestAccDataSourceAWSS3BucketObjects_prefixes (117.61s)
--- PASS: TestAccDataSourceAWSS3BucketObjects_startAfter (102.55s)

--- PASS: TestAccDataSourceAwsSecurityGroups_filter (70.25s)
--- PASS: TestAccDataSourceAwsSecurityGroups_tag (68.53s)

--- PASS: TestAccDataSourceAwsVpcPeeringConnections_basic (70.67s)

--- PASS: TestAccDataSourceAwsVpcs_basic (65.12s)
--- PASS: TestAccDataSourceAwsVpcs_filters (56.35s)
--- PASS: TestAccDataSourceAwsVpcs_tags (58.15s)

--- SKIP: TestAccAWSOutpostsOutpostsDataSource_basic (2.12s)

--- SKIP: TestAccAWSOutpostsSitesDataSource_basic (1.42s)

--- SKIP: TestAccDataSourceAwsEc2CoipPools_basic (15.89s)
--- SKIP: TestAccDataSourceAwsEc2CoipPools_Filter (14.35s)

--- SKIP: TestAccDataSourceAwsEc2LocalGatewayRouteTables_basic (14.72s)
--- SKIP: TestAccDataSourceAwsEc2LocalGatewayRouteTables_Filter (2.24s)

--- SKIP: TestAccDataSourceAwsEc2LocalGateways_basic (2.51s)

--- SKIP: TestAccDataSourceAwsEc2LocalGatewayVirtualInterfaceGroups_basic (3.57s)
--- SKIP: TestAccDataSourceAwsEc2LocalGatewayVirtualInterfaceGroups_Filter (2.13s)
--- SKIP: TestAccDataSourceAwsEc2LocalGatewayVirtualInterfaceGroups_Tags (1.99s)
```

### CODE REVIEW

File: aws/data_source_aws_kms_ciphertext.go [hunk 1]
```diff
@@ -60,6 +57,8 @@ func dataSourceAwsKmsCiphertextRead(d *schema.ResourceData, meta interface{}) er
 		return err
 	}
 
+	d.SetId(d.Get("key_id").(string))
```
Comment (anGie44) [inline aws/data_source_aws_kms_ciphertext.go:60]: could we use `resp.KeyId` here?

File: website/docs/d/kms_ciphertext.html.markdown [hunk 2]
```diff
@@ -48,4 +48,5 @@ The following arguments are supported:
 
 All of the argument attributes are also exported as result attributes.
 
+* `id` - AWS Region.
```
Comment (anGie44) [inline website/docs/d/kms_ciphertext.html.markdown:51]:
```suggestion
* `id` - Globally unique key ID for the customer master key.
```

**Review by anGie44 — APPROVED** [review]
LGTM 🙌 just small comments re: the kms data source ID change

File: aws/data_source_aws_kms_ciphertext.go (hunk 1 above)
Comment (bflad) [reply aws/data_source_aws_kms_ciphertext.go:60]: Sure can!

```suggestion
	d.SetId(aws.StringValue(resp.KeyId))
```

**Comment by bflad** [comment]
Verified acceptance testing after `aws_kms_ciphertext` data source update:

```
--- PASS: TestAccDataSourceAwsKmsCiphertext_validate (17.86s)
--- PASS: TestAccDataSourceAwsKmsCiphertext_basic (18.17s)
--- PASS: TestAccDataSourceAwsKmsCiphertext_validate_withContext (18.65s)
```

**Pull request merged**